src/GovUk.Frontend.AspNetCore/TagHelpers/CharacterCountTagHelper.cs

```diff
@@ -254,8 +252,32 @@ private protected override IHtmlContent GenerateFormGroupContent(
         var textAreaTagBuilder = GenerateTextArea(haveError);
         contentBuilder.AppendHtml(textAreaTagBuilder);
 
+        var countRemainingMessage = GenerateCountRemainingMessage();
+        contentBuilder.AppendHtml(countRemainingMessage);
+
         return contentBuilder;
 
+
+        IHtmlContent GenerateCountRemainingMessage()
+        {
+            // TODO / FIXME: temporary hardcoded, need to pass through from somewhere? Unable to access context from here?
+            string textAreaDescriptionText = null;
+            var hasNoLimit = MaxLength is null && MaxWords is null; // TODO: This logic is reused in many places -- centralise it?
+
+            var hintId = $"{ResolveIdPrefix()}-info";
+
+            var content = hasNoLimit ? "" :
+                (textAreaDescriptionText ?? $"You can enter up to %{{count}} {(MaxWords.HasValue ? "words" : "characters")}")
+                .Replace("%{count}", (MaxWords.HasValue ? MaxWords : MaxLength).ToString());
+
+            var hintContent = new HtmlString(HtmlEncoder.Default.Encode(content));
+
+            var attributes = CountMessageAttributes.ToAttributeDictionary();
+            attributes.MergeCssClass("govuk-character-count__message");
+
+            return Generator.GenerateHint(hintId, hintContent, attributes);
+        }
```

Comment:
Flagging the to-do here. It's one of the examples where there seems to be a mismatch between data being stored on the `*Context` but isn't available on the `FormGroupTagHelperBase#GenerateFormGroupContent` method / overrides.

Note that tests aren't failing with this dodgy implementation, but maybe this means additional tests are needed?